### PR TITLE
feat: foundations bundle — operators, viz utilities, metric helpers

### LINKED ISSUE #116: viz: SpatialMapPanel — single 2-D field snapshot with optional cartopy
## Problem / Request

Add `SpatialMapPanel` for single-snapshot 2-D field rendering — the missing baseline for any benchmarking notebook that needs to show "what the field looks like".

## User Story

> As a user comparing predictions to a reference, I want a one-line panel call that draws a labelled map of a 2-D `(lat, lon)` snapshot, so that I don't reach for raw `pcolormesh` boilerplate in every notebook.

## Motivation

Inspired by `oceanbench/notebooks/dev/neurips-paper/utils.py::plot_sst_map`. Currently every demo notebook (V2, V4, ocean tutorials, V1.5) hand-rolls a `plt.subplots(...)` + `xr.DataArray.plot()` cell to show a snapshot. Promoting this to a panel:

- Keeps `savefig` / `show` parity with our other panels.
- Standardises diverging-cmap defaults, vmin/vmax, gridlines, optional cartopy projection.
- Slots into Sequential / Graph as a terminal viz step.

## Proposed API

```python
class SpatialMapPanel(_ValidationPanel):
    """Single 2-D field snapshot with optional cartopy projection."""

    def __init__(
        self,
        *,
        var: str | None = None,
        time_index: int | None = 0,
        cmap: str = "RdBu_r",
        vmin: float | None = None,
        vmax: float | None = None,
        projection: str | None = None,   # "PlateCarree", "Robinson", ...
        coastlines: bool = False,
        gridlines: bool = False,
        cbar_label: str = "",
        **kw,
    ) -> None: ...
```

## Design Snapshot

- Input: `xr.DataArray` or single-var `xr.Dataset` with `(lat, lon)` and optional leading `time`.
- When `projection=None`: plain matplotlib axes (no cartopy dep).
- When `projection="PlateCarree"`: optional cartopy import; raise a clear `ImportError` if not installed.
- Reuse `_ValidationPanel` plumbing for figsize / title / savefig / show.

## Acceptance

- Renders a Figure for a `(lat, lon, [time])` DataArray.
- Test asserts savefig path created.
- Cartopy path is opt-in (no hard dependency).
- Wired into `docs/api/validation.md`.

### LINKED ISSUE #119: viz: method_palette() helper for consistent multi-method colours
## Problem / Request

Add a small `method_palette()` helper (and `palette` kwarg on multi-method panels) so colours stay consistent across iso PSD, space-time PSD, and time-series figures.

## User Story

> As a user producing a multi-figure benchmark report, I want method colours to stay identical across all panels so that the reader can track "DUACS = green" through every figure without me hand-coding `tab:green` repeatedly.

## Motivation

`oceanbench/notebooks/dev/neurips-paper/utils.py` hard-codes a per-method palette. We can promote this to a small reusable helper so all our panels (PSD iso/score/space-time + TimeSeriesErrorPanel) share a deterministic mapping.

## Proposed API

```python
from xr_toolz.viz.validation import method_palette

# Auto-assign deterministic colours.
palette = method_palette(["DUACS", "MIOST", "BFNQG", "4DVarNet"])
# {"DUACS": "tab:green", ...}

# Or pass into a panel directly.
PSDIsotropicPanel(palette=palette)(iso_ds)
TimeSeriesErrorPanel(palette=palette)(scores_ds)
```

## Acceptance

- Pure function, returns `dict[str, str]`.
- Deterministic order for any iterable of names (sorted, then cycled through `tab:*` colours).
- `palette` kwarg accepted on the multi-method panels (PSDIsotropicPanel + TimeSeriesErrorPanel + future ones).
- 100-line module + 5-line test.

### LINKED ISSUE #121: geo: FillNaN Operator — constant-fill (promote inline V1.5 op)
## Problem / Request

Promote the inline `FillNaN` Operator from `docs/notebooks/validation_v1_psd_plots.ipynb` into the library so prep pipelines don't need to define it ad hoc.

## User Story

> As a user building a Sequential prep pipeline, I want an `FillNaN(value=0.0)` operator already in `xr_toolz` so that I don't have to write a 10-line inline subclass in every notebook that needs to zero out land cells before PSD scoring.

## Motivation

Came up while wiring the V1.5 PSD demo's bonus end-to-end Graph DAG. `xr_toolz.interpolate.FillNaNSpatial / FillNaNTemporal` exist but they *interpolate* across NaNs — there's no constant-fill op. PSD prep needs zero-fill (land cells = no signal); a generic `FillNaN(value)` is the cleanest fix.

## Proposed API

```python
# src/xr_toolz/geo/operators.py
class FillNaN(Operator):
    """Replace NaN with a constant value."""

    def __init__(self, value: float = 0.0) -> None:
        self.value = float(value)

    def _apply(self, ds: xr.Dataset) -> xr.Dataset:
        return ds.fillna(self.value)

    def get_config(self) -> dict[str, Any]:
        return {"value": self.value}
```

## Acceptance

- Lives in `xr_toolz.geo.operators` (alongside `RemoveMean`).
- Re-exported via `xr_toolz.geo`.
- Docstring + 2 tests (default 0.0, custom value).
- Used in `validation_v1_psd_plots.ipynb` (replace inline class definition).

### LINKED ISSUE #122: geo: Reduce Operator — generic dim reduction (mean/sum/median/...)
## Problem / Request

Add a generic `Reduce` Operator that wraps `xr.Dataset.mean / sum / median / max / min` over an arbitrary dim. Promotes the inline `MeanOverDim` from the V1.5 PSD demo notebook.

## User Story

> As a user composing a Graph DAG that ends with "average over time before plotting", I want a library `Reduce("mean", "time")` op so that I don't sprinkle one-off `MeanOverDim`-style classes in every notebook.

## Motivation

Came up in the V1.5 PSD bonus DAG cell. Generic dim-reduction is the most common composition gap — current library lacks any reducer Operator. Pairs with `FillNaN` (#prev).

## Proposed API

```python
# src/xr_toolz/geo/operators.py  (or transforms — TBD by reviewer)
class Reduce(Operator):
    """Reduce a Dataset over one or more dims.

    Args:
        op: One of "mean" / "sum" / "median" / "min" / "max" / "std".
        dim: Dim or tuple of dims to reduce over.
        keepdims: Whether to keep reduced dims as length-1.
    """

    def __init__(
        self,
        op: str = "mean",
        dim: str | tuple[str, ...] = ("time",),
        *,
        keepdims: bool = False,
    ) -> None: ...
```

## Acceptance

- Validate `op` against an allowlist; clear error on unknown.
- Tests: each op produces the expected result vs xarray equivalent.
- Replace inline `MeanOverDim` in `validation_v1_psd_plots.ipynb`.

### LINKED ISSUE #123: viz: shared_norm() utility for matched (vmin, vmax) across panels
## Problem / Request

Add a `shared_norm()` utility returning matched `(vmin, vmax)` across multiple DataArrays, for fair multi-panel comparison plots.

## User Story

> As a user producing a 4-method comparison grid, I want a one-liner that picks a single `(vmin, vmax)` (e.g. the 2nd / 98th percentiles across all inputs) so that the colour scale is identical on every subplot — otherwise the eye sees structural differences that are pure cmap-stretch artifacts.

## Motivation

Used implicitly in oceanbench's NeurIPS paper grid figures. Trivial to implement (≤20 lines), but absence forces every comparison notebook to hand-roll percentile reductions.

## Proposed API

```python
from xr_toolz.viz import shared_norm

vmin, vmax = shared_norm(da_a, da_b, da_c, q=(0.02, 0.98), symmetric=False)

# Or for divergent fields:
vmin, vmax = shared_norm(*errors, symmetric=True)   # → (-M, +M) where M = max(|q_low|, q_high)
```

## Acceptance

- Accepts variadic DataArrays / Datasets (single var auto-extracted).
- Quantile-based by default with `q=(0.02, 0.98)`; full-range opt-in via `q=None`.
- `symmetric=True` returns symmetric limits around zero.
- ≤25 lines + 4 tests.

### LINKED ISSUE #124: viz: clip toggle on PSD score panels for signed-score diagnostic mode
## Problem / Request

Add a `clip: bool = True` toggle on `PSDIsotropicScorePanel` and `PSDSpaceTimeScorePanel` so users can render the *signed* (un-clipped) score for diagnostic deep-dives.

## User Story

> As a user investigating *why* a score is poor, I want to see negative score values explicitly (where `PSD(error) > PSD(ref)`) so that I can distinguish "slightly worse than zero predictor" from "catastrophically wrong".

## Motivation

Currently both score panels clip to `[0, 1]` to match the oceanbench / Ballarotta convention. That's the right *default*, but the un-clipped score is informative for debugging and for showing how a baseline degrades.

## Proposed API

```python
panel = PSDIsotropicScorePanel(
    threshold=0.5,
    clip=False,          # NEW — show signed score
    ylim=(-1.5, 1.0),    # explicit ylim helpful when clip=False
)
```

## Acceptance

- `clip=True` (default) preserves current behaviour — full backwards compat.
- `clip=False`: ylim auto-computed from data unless overridden; negative values rendered.
- Add a test asserting both modes produce a Figure for a fixture with negative scores.
- Update docstring with the convention rationale.

### LINKED ISSUE #126: metrics: rank_methods() leaderboard DataFrame builder
## Problem / Request

Add a `rank_methods()` helper producing a sorted leaderboard DataFrame from a `(method, metric)` Dataset.

## User Story

> As a user writing a benchmark report, I want a function that takes a Dataset of `(method, region)` scores and returns a sorted-by-metric DataFrame so that I can drop a clean Markdown / LaTeX table into the paper without hand-formatting it.

## Motivation

Standard output of every benchmarking notebook (oceanbench paper, SSH challenge, etc.). Pairs with `TimeSeriesErrorPanel` to give both visual and tabular summaries from the same source.

## Proposed API

```python
from xr_toolz.metrics import rank_methods

# scores: Dataset with data_vars = metrics, dim = "method"
ranked = rank_methods(
    scores,
    by="rmse",                   # primary sort key
    ascending=True,              # lower is better
    include=("rmse", "nrmse"),   # subset of metrics to keep in output
    region_dim=None,             # or "region" for per-region tables
)
print(ranked.to_markdown())  # paper-ready
```

## Acceptance

- Returns a `pandas.DataFrame` indexed by method (and optionally region).
- Stable sort, ties broken by next column.
- Test against synthetic 3-method × 3-metric Dataset.
- Documented with `.to_markdown()` rendering example.

### LINKED ISSUE #127: metrics: find_intercept_2D — extract score-threshold contour as data
## Problem / Request

Add `find_intercept_2D()` — locate the 2-D contour where a `(freq_lon, freq_time)` score crosses a threshold, returning the boundary as a polyline DataArray.

## User Story

> As a user reporting space-time effective resolution, I want to extract the *resolved boundary* in `(k, ω)` space as data (not just a contour line on a plot) so that I can quote it as a number in a paper or compare boundaries across methods quantitatively.

## Motivation

Counterpart to existing `find_intercept_1D`. oceanbench uses the 2-D variant for space-time score boundaries; we currently only overlay it as a `contour` line on the `PSDSpaceTimeScorePanel` figure — there's no programmatic accessor to the boundary itself.

## Proposed API

```python
from xr_toolz.metrics import find_intercept_2D

boundary = find_intercept_2D(
    score,                     # (freq_lon, freq_time) DataArray
    level: float = 0.5,
    space_dim="freq_lon",
    time_dim="freq_time",
)
# Returns a DataArray with dims ("segment", "point", "axis") or similar —
# review final shape during implementation.
```

## Acceptance

- Wraps `skimage.measure.find_contours` or matplotlib's `contour` extraction.
- Handles disconnected segments cleanly (returns a list / variable-length contours).
- Test: synthetic monotone score → contour matches analytic boundary to within 1 cell.
- Pairs with `PSDSpaceTimeScorePanel.boundary_data` accessor (optional follow-up).

### LINKED ISSUE #128: viz: cartopy projection presets (gulf_stream, north_atlantic, kuroshio, ...)
## Problem / Request

Add cartopy projection presets (Gulf Stream, North Atlantic, Global, Kuroshio, Med) usable by `SpatialMapPanel` / `SpatialDiffPanel`.

## User Story

> As a user plotting a Gulf Stream snapshot, I want `projection="gulf_stream"` to apply the right `set_extent`, projection, and feature scales so that I don't memorise lat/lon bounds for every basin.

## Motivation

Standardises basin-specific cartopy boilerplate that every ocean-paper notebook reproduces. Cartopy stays optional — presets are only resolved when a user actually requests one.

## Proposed API

```python
# xr_toolz.viz.projections
PRESETS = {
    "global":      {"projection": "Robinson",     "extent": None},
    "north_atlantic": {"projection": "PlateCarree", "extent": (-80, 0, 10, 65)},
    "gulf_stream": {"projection": "PlateCarree", "extent": (-80, -50, 30, 45)},
    "kuroshio":    {"projection": "PlateCarree", "extent": (130, 180, 25, 45)},
    "mediterranean": {"projection": "PlateCarree", "extent": (-6, 36, 30, 46)},
}

def make_axes(projection: str | None = None, **kw): ...
```

## Acceptance

- Cartopy import only fires when a preset is requested (lazy).
- `SpatialMapPanel(projection="gulf_stream")` works end-to-end.
- Tests skip cleanly if cartopy isn't installed.
- Document the `PRESETS` dict in the API page.

## Dependencies

- Adds an *optional* extras group `[viz-geo]` with `cartopy >= 0.22`.
- Hard-fails with a clear `ImportError("install xr_toolz[viz-geo]")` if used without the extra.

### PR DESCRIPTION
**Pull request by jejjohnson**
## Summary

Bundles small follow-up issues from the V1.5 PSD work into one reviewable PR — operators / metric helpers / viz utilities + the spatial-snapshot panel with cartopy presets.

Closes #116
Closes #119
Closes #121
Closes #122
Closes #123
Closes #124
Closes #126
Closes #127
Closes #128

### geo
- **#121 `FillNaN(value=0.0)`** — constant-fill Operator. Counterpart to the existing interpolating `FillNaNSpatial` / `FillNaNTemporal` for zeroing land cells before PSD scoring.
- **#122 `Reduce(op, dim, keepdims=False)`** — generic dim reducer wrapping `mean / sum / median / min / max / std / var`.

### viz
- **#123 `shared_norm(*arrays, q=(0.02, 0.98), symmetric=False)`** — matched `(vmin, vmax)` across inputs for fair comparison grids.
- **#119 `viz.validation.method_palette(names, cycle=…)`** — deterministic `{name: colour}` mapping for consistent multi-method colours.
- **#124 `clip` toggle** on `PSDIsotropicScorePanel` / `PSDSpaceTimeScorePanel` (plus `ylim` on the iso variant). Default `clip=True` preserves current behaviour.
- **#128 cartopy projection presets** — `xr_toolz.viz.PRESETS` (global / north_atlantic / gulf_stream / kuroshio / mediterranean) + `make_axes()`. Cartopy promoted to a core dep.
- **#116 `SpatialMapPanel`** — single 2-D `(lat, lon)` snapshot panel inheriting `_ValidationPanel`. Auto-cmap from the registry; optional cartopy preset; `time_index` selection; Dataset auto-pick.

### metrics
- **#127 `find_intercept_2D(score, level, space_dim, time_dim)`** — 2-D counterpart to `find_intercept_1D`. Returns the score-threshold contour as a list of polyline DataArrays.
- **#126 `rank_methods(scores, by, …)`** — sorted leaderboard DataFrame builder for `(method, metric)` Datasets with optional `region_dim`.

### types
- `Variable` gains a `cmap` field; the curated `REGISTRY` carries a field-appropriate matplotlib colormap on every entry (SSH → `RdBu_r`, SST → `RdYlBu_r`, salinity → `viridis`, etc.).
- New "Ocean kinematic diagnostics" registry section: 19 derived-quantity entries (`psi`, `ke`, `eke`, `speed`, `ens`, `vort_r/_a/_shear/_curv`, `div`, `shear_strain`, `tensor_strain`, `strain`, `ow`, `u_a`, `v_a`, `n_squared`, `lapse_rate`, `mld`, `pv_barotropic`) so panels and helpers can resolve sensible cmaps for diagnostics, not just CF physical variables.
- `xr_toolz.viz.cmap_for(var)` resolves the registry entry's cmap (case-insensitive on string names; accepts `Variable` instances).

### docs / notebook
- `docs/api/validation.md` adds a Spatial-snapshots section documenting `SpatialMapPanel`, the cartopy presets, and the variable→cmap registry.
- `docs/notebooks/ocean_kinematics_glorys_duacs.ipynb` opens with a `SpatialMapPanel(var="ssh", projection="gulf_stream")` snapshot, and every `two_panel` call site now passes `variable=` instead of hard-coded `cmap=` — colours auto-resolve from the registry. Re-executed end-to-end.

## Test plan

- [x] 23 new tests in `tests/test_foundations_bundle.py` cover the foundations helpers + clip toggle.
- [x] 6 new tests in `tests/test_geo_operators.py` cover `FillNaN` and `Reduce`.
- [x] 34 new tests in `tests/test_viz_spatial.py` cover `Variable.cmap` per family, `cmap_for` resolution, `PRESETS` / `make_axes` plain vs cartopy, `SpatialMapPanel` auto-cmap / explicit-cmap / no-time / Dataset auto-pick / preset extent / savefig.
- [x] Full suite: **1017 passed**, 1 skipped.
- [x] `ruff check .` / `ruff format --check .` / `ty check src/xr_toolz` clean (typecheck count unchanged from main).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/xr_toolz/viz/validation/_src/psd.py [hunk 1]
```diff
@@ -273,6 +285,8 @@ def __init__(
         # from the trailing ``[unit]`` of ``wavelength_label`` (so
         # ``"Wavelength [km]"`` → ``"km"``); falls back to empty.
         self.resolved_units = resolved_units
+        self.clip = bool(clip)
+        self.ylim = tuple(ylim) if ylim is not None else None
```
[the remark this answers is not in the document]
Comment (jejjohnson) [reply src/xr_toolz/viz/validation/_src/psd.py]: Fixed in e9d8f87 — `__init__` now validates that `ylim` is a 2-tuple with `ymin <= ymax` and raises a clear `ValueError` otherwise.

File: src/xr_toolz/viz/_src/norm.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                raise ValueError(
+                    "shared_norm: Dataset inputs must have exactly one "
+                    f"data variable; got {list(a.data_vars)}."
+                )
+            (var,) = a.data_vars
+            arr = a[var].values
+        else:
+            arr = a.values
+        flats.append(np.asarray(arr).ravel())
+    flat = np.concatenate(flats)
```
[the remark this answers is not in the document]
Comment (jejjohnson) [reply src/xr_toolz/viz/_src/norm.py]: Fixed in e9d8f87 — `shared_norm` now reduces through xarray (`xr.concat` + `.quantile` / `.min` / `.max`) on a stacked flat dim, so dask-backed inputs stay lazy until the final scalar is realised. Added a `pytest.importorskip('dask.array')`-guarded test.

File: src/xr_toolz/metrics/_src/spectral.py [hunk 3]
```diff
@@ -204,6 +204,77 @@ def find_intercept_1D(
...
+        List of :class:`xr.DataArray`, one per contour segment, each
+        with dims ``("point", "axis")`` where ``axis`` has coordinates
+        ``(space_dim, time_dim)`` and contains the segment's coordinate
+        polyline.
+    """
+    try:
+        from skimage.measure import find_contours
+    except ImportError as exc:  # pragma: no cover - optional dep
+        raise ImportError(
+            "find_intercept_2D requires scikit-image. Install with "
```
[the remark this answers is not in the document]
Comment (jejjohnson) [reply src/xr_toolz/metrics/_src/spectral.py]: Fixed in e9d8f87 — dropped the `try/except ImportError` guard. `scikit-image` is already in the package's required `dependencies` and other metrics modules import it unguarded; the conditional was misleading.

File: src/xr_toolz/metrics/_src/leaderboard.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        # Sort within each region by `by`, ties broken by remaining cols.
+        sort_cols = [by, *[c for c in keep if c != by]]
+        df = df.groupby(level=region_dim, group_keys=False).apply(
+            lambda g: g.sort_values(sort_cols, ascending=ascending, kind="stable")
+        )
+    else:
+        if df.index.nlevels > 1:
+            # Squeeze any singleton non-method levels.
+            extra = [n for n in df.index.names if n != method_dim]
+            df = df.reset_index(extra, drop=True)
```
[the remark this answers is not in the document]
Comment (jejjohnson) [reply src/xr_toolz/metrics/_src/leaderboard.py:86]: Fixed in e9d8f87 — when `region_dim` is `None`, the squeeze path now rejects multi-valued extra index levels with a clear `ValueError` ("multi-valued extra index level(s) ... besides 'method'; pass `region_dim=...`") instead of silently dropping them. Singleton extras are still squeezed silently. Added a test that asserts a forgotten `region_dim` on a `(region, method)` Dataset raises.
